File: 848.字母移位.js
/*
 * @lc app=leetcode.cn id=848 lang=javascript
 *
 * [848] 字母移位
 */

// @lc code=start
/**
 * @param {string} s
 * @param {number[]} shifts
 * @return {string}
 */
var shiftingLetters = function(s, shifts) {
    const n = s.length;
    const result = new Array(n);
    let totalShift = 0;

    for (let i = n - 1; i >= 0; i--) {
        totalShift += shifts[i];
        result[i] = String.fromCharCode(((s.charCodeAt(i) - 'a'.charCodeAt(0) + totalShift) % 26) + 'a'.charCodeAt(0));
    }

    return result.join('');
};
// @lc code=end

File: 848.字母移位.java
/*
 * @lc app=leetcode.cn id=848 lang=java
 *
 * [848] 字母移位
 */

// @lc code=start
class Solution {
    public String shiftingLetters(String s, int[] shifts) {
        int n = s.length();
        char[] result = new char[n];
        long totalShift = 0;

        // 从后往前计算每个字符需要移位多少
        for (int i = n - 1; i >= 0; i--) {
            totalShift += shifts[i];
            // 移位
            result[i] = (char) ((s.charAt(i) - 'a' + totalShift) % 26 + 'a');
        }

        return new String(result);
    }
}
// @lc code=end

